components/fxa-client/src/jwcrypto.rs

```diff
@@ -0,0 +1,265 @@
+}
+#[derive(Serialize, Deserialize, PartialEq)]
+pub enum EncryptionAlgorithm {
+    A256GCM,
+}
+
+impl EncryptionAlgorithm {
+    fn algorithm_id(&self) -> &'static str {
+        match self {
+            Self::A256GCM => "A256GCM",
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct JweHeader {
+    pub alg: Algorithm,
+    pub enc: EncryptionAlgorithm,
+    pub epk: Option<Jwk>,
+    pub apu: Option<String>,
+    pub apv: Option<String>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Jwk {
+    pub crv: String,
+    pub kty: String,
+    pub x: String,
+    pub y: String,
+}
+
+/// Encrypt and serialize data in the JWE compact form.
+pub fn encrypt_to_jwe(data: &[u8], encryption_params: EncryptionParameters) -> Result<String> {
+    let (encrypted_key, protected_header, iv, ciphertext, auth_tag) = match encryption_params {
+        EncryptionParameters::ECDH_ES {
+            enc,
+            local_key_pair,
+            peer_jwk,
+        } => {
+            let local_public_key = extract_pub_key_jwk(&local_key_pair)?;
+            let protected_header = JweHeader {
+                alg: Algorithm::ECDH_ES,
+                enc,
+                epk: Some(local_public_key),
+                apu: None,
+                apv: None,
+            };
+
+            let secret = derive_shared_secret(&protected_header, local_key_pair, &peer_jwk)?;
+            let encryption_algorithm = match protected_header.enc {
+                EncryptionAlgorithm::A256GCM => &aead::AES_256_GCM,
+            };
+            let sealing_key = aead::SealingKey::new(encryption_algorithm, &secret.as_ref())?;
+            let additional_data = serde_json::to_string(&protected_header)?;
+            let additional_data =
+                base64::encode_config(additional_data.as_bytes(), base64::URL_SAFE_NO_PAD);
+            let additional_data = additional_data.as_bytes();
+            let aad = aead::Aad::from(additional_data);
+            let mut iv: Vec<u8> = vec![0; 12];
+            rand::fill(&mut iv)?;
+            let nonce = aead::Nonce::try_assume_unique_for_key(encryption_algorithm, &iv)?;
+            let mut encrypted = aead::seal(&sealing_key, nonce, aad, data)?;
+
+            let tag_idx = encrypted.len() - ((128 + 7) >> 3);
+            let tag = encrypted.split_off(tag_idx);
+            let ciphertext = encrypted;
+
+            (None, protected_header, iv, ciphertext, tag)
+        }
+    };
+
+    // Assemble JWE
+    let encrypted_key = encrypted_key
+        .map(|k: Vec<u8>| base64::encode_config(&k, base64::URL_SAFE_NO_PAD))
+        .unwrap_or_default();
+    let header = serde_json::to_string(&protected_header)?;
+    let header = base64::encode_config(header.as_bytes(), base64::URL_SAFE_NO_PAD);
+    let tag = base64::encode_config(auth_tag, base64::URL_SAFE_NO_PAD);
+    let ciphertext = base64::encode_config(ciphertext, base64::URL_SAFE_NO_PAD);
+    let iv = base64::encode_config(iv, base64::URL_SAFE_NO_PAD);
```

Comment:
pico-nit: I was momentarily alarmed when you encoded the tag before the ciphertext, but I see you're joining them together in the correct order below. It may be worth b64-encoding these items in the same order in which they'll appear in the final JWE, for a tiny bit less mental load when scanning this code.